src/DynamoCore/Scheduler/UpdateRenderPackageAsyncTask.cs

```diff
@@ -212,104 +280,173 @@ protected override void HandleTaskExecutionCore()
+
+            package.AddTriangleVertexUV(0, 0);
+            package.AddTriangleVertexUV(0, 0);
+            package.AddTriangleVertexUV(0, 0);
+            package.AddTriangleVertexUV(0, 0);
+            package.AddTriangleVertexUV(0, 0);
+            package.AddTriangleVertexUV(0, 0);
+
+            // Draw plane edges
+            package.AddLineStripVertex(a.X, a.Y, a.Z);
+            package.AddLineStripVertex(b.X, b.Y, b.Z);
+            package.AddLineStripVertex(b.X, b.Y, b.Z);
+            package.AddLineStripVertex(c.X, c.Y, c.Z);
+            package.AddLineStripVertex(c.X, c.Y, c.Z);
+            package.AddLineStripVertex(d.X, d.Y, d.Z);
+            package.AddLineStripVertex(d.X, d.Y, d.Z);
+            package.AddLineStripVertex(a.X, a.Y, a.Z);
+
+            // Draw normal
+            package.AddLineStripVertex(plane.Origin.X, plane.Origin.Y, plane.Origin.Z);
+            var nEnd = plane.Origin.Add(plane.Normal.Scale(2.5));
+            package.AddLineStripVertex(nEnd.X, nEnd.Y, nEnd.Z);
+
+            for (var i = 0; i < 5; i++)
+            {
+                package.AddLineStripVertexCount(2);
+                package.AddLineStripVertexColor(MidTone, MidTone, MidTone, 255);
+                package.AddLineStripVertexColor(MidTone, MidTone, MidTone, 255);
+            }
 
-                        for (var i = 0; i < package.MeshVertexCount; i++)
-                        {
-                            package.AddTriangleVertexNormal(plane.Normal.X, plane.Normal.Y, plane.Normal.Z);
-                        }
+            for (var i = 0; i < 6; i++)
+            {
+                package.AddTriangleVertexNormal(plane.Normal.X, plane.Normal.Y, plane.Normal.Z);
+                package.AddTriangleVertexColor(0, 0, 0, 10);
+            }
+        }
 
-                        for (var i = 0; i < package.MeshVertexCount; i++)
-                        {
-                            package.AddTriangleVertexColor(0, 0, 0, 10);
-                        }
-                    }
+        private static void EnsureColorExistsPerVertex(IRenderPackage package, int previousPointVertexCount, int previousLineVertexCount, int previousMeshVertexCount)
+        {
+            var packageSupplement = package as IRenderPackageSupplement;
+            
+            //Use legacy slow path if package does not implement IRenderPackageSupplement
+            if (packageSupplement == null)
+            {
+                LegacyPackageEnsureColorExistsPerVertex(package, previousPointVertexCount, previousLineVertexCount, previousMeshVertexCount);
+                return;
+            }
+            
+            if (package.PointVertexCount > previousPointVertexCount)
+            {
+                var count = package.PointVertexCount - packageSupplement.PointVertexColorCount;
+                if (count > 0)
+                {
+                    packageSupplement.AppendPointVertexColorRange(CreateColorByteArrayOfSize(count, DefR, DefG, DefB, DefA));
```

Comment:
How is this different from `AddPointVertexColor` called in a for loop?

Reply:
functionally there isn't a huge difference... under the hood it uses `AddRange`.  I think it just is clearer code.